src/pagination.rs

```diff
@@ -0,0 +1,312 @@
+        );
+        object.insert(
+            "last_index_permalink".into(),
+            liquid::value::Value::scalar(self.last_index_permalink),
+        );
+        object.insert(
+            "total_indexes".into(),
+            liquid::value::Value::scalar(self.total_indexes as i32),
+        );
+        object.insert(
+            "total_pages".into(),
+            liquid::value::Value::scalar(self.total_pages as i32),
+        );
+        object
+    }
+}
+
+pub fn generate_paginators(
+    doc: &mut Document,
+    posts_data: &[liquid::value::Value],
+) -> Result<Vec<Paginator>> {
+    let config = doc.front.pagination.as_ref().unwrap();
+    let mut all_posts = posts_data.to_vec();
+    match config.include {
+        Include::All => {
+            sort_posts(&mut all_posts, &config);
+            create_all_paginators(&mut all_posts, &doc, &config)
+        }
+        Include::None => Ok(Vec::new()), // user can set `include: None`
+    }
+}
+
+fn extract_value(a: &liquid::value::Value, key: &str) -> Option<liquid::value::Scalar> {
+    a.get(&key.into())
+        .and_then(|sort_key| sort_key.as_scalar().cloned())
+}
+
+// sort posts by multiple criteria
+fn sort_posts(posts: &mut Vec<liquid::value::Value>, config: &PaginationConfig) {
+    // Boxing needed to compile, don't really understand why
+    let order: Box<Fn(liquid::value::Scalar, liquid::value::Scalar) -> Ordering> = match config
+        .order
+    {
+        SortOrder::Desc => {
+            Box::new(|a, b: liquid::value::Scalar| b.partial_cmp(&a).unwrap_or(Ordering::Equal))
+        }
+        SortOrder::Asc => {
+            Box::new(|a: liquid::value::Scalar, b| a.partial_cmp(&b).unwrap_or(Ordering::Equal))
+        }
+        SortOrder::None => {
+            // when built, order is set like this:
+            // `order.unwrap_or(SortOrder::Desc);` so it's unreachable
+            unreachable!("Sort order should have default value when constructing PaginationConfig")
+        }
+    };
+    posts.sort_by(|a, b| {
+        let keys = &config.sort_by;
+        let mut cmp = Ordering::Less;
+        for k in keys {
+            cmp = match (extract_value(a, &k), extract_value(b, &k)) {
+                (Some(a), Some(b)) => order(a, b),
+                (None, None) => Ordering::Equal,
+                (_, None) => Ordering::Greater,
+                (None, _) => Ordering::Less,
+            };
+            if cmp != Ordering::Equal {
+                return cmp;
+            }
+        }
+        cmp
+    });
+}
+
+fn create_all_paginators(
+    mut all_posts: &mut Vec<liquid::value::Value>,
+    doc: &Document,
+    pagination_cfg: &PaginationConfig,
+) -> Result<Vec<Paginator>> {
+    let total_pages = all_posts.len();
+    // f32 used here in order to not lose information to ceil the result,
+    // otherwise we can lose an index
+    let total_indexes = (total_pages as f32 / pagination_cfg.per_page as f32).ceil() as usize;
```

Comment:
Alternatively, you could use div/rem or [`div_rev`](https://rust-num.github.io/num/num/trait.Integer.html#tymethod.div_rem)

Reply:
How so? 
```rust
assert_eq!(( 8).div_rem( &3), ( 2,  2));
```
when in this case I want 3 as quotient.

`div_ceil` would be what I want, but it does not exist.

Reply:
No offence taken :) I know you're way better than me so I tend to learn from you :D I also feel that casting is a bit overkill. I though you got a function unknown from me to do that in a one-liner. Maybe it's more efficient to test the remainder, so we stay out of floating type?

Reply:
For now, its fine to stay with float.